hub/core/dataset/dataset.py

```diff
@@ -96,31 +98,37 @@ def __init__(
             AuthorizationException: If a Hub cloud path (path starting with hub://) is specified and the user doesn't have access to the dataset.
             PathNotEmptyException: If the path to the dataset doesn't contain a Hub dataset and is also not empty.
         """
+        d: Dict[str, Any] = {}
+        d["_client"] = d["org_id"] = d["ds_name"] = None
         # uniquely identifies dataset
-        self.path = path or get_path_from_storage(storage)
-        self.storage = storage
-        self._read_only = read_only
-        base_storage = get_base_storage(storage)
-        if (
-            not read_only and index is None and isinstance(base_storage, S3Provider)
-        ):  # Dataset locking only for S3 datasets
-            try:
-                lock(base_storage, callback=lambda: self._lock_lost_handler)
-            except LockedException:
-                self.read_only = True
-                warnings.warn(
-                    "Opening dataset in read only mode as another machine has locked it for writing."
-                )
-
-        self.index: Index = index or Index()
-        self.group_index = group_index
-        self._token = token
-        self.public = public
-        self.verbose = verbose
-        self.is_first_load = version_state is None
-        self.version_state: Dict[str, Any] = version_state or {}
-        self._info = None
+        d["path"] = path or get_path_from_storage(storage)
```

Comment:
using dict assignments to avoid our custom setattr which slows down object creation substantially